infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,109 @@
+package org.corfudb.infrastructure.logreplication;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationEvent;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationState;
+import org.corfudb.infrastructure.logreplication.replication.fsm.LogReplicationStateType;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.IllegalTransitionException;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeEvent;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeState;
+import org.corfudb.infrastructure.logreplication.runtime.fsm.LogReplicationRuntimeStateType;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+
+@Slf4j
+public class FsmTaskManager {
+
+    private ExecutorService fsmWorker;
+
+    public FsmTaskManager (String threadName, int threadCount) {
+        fsmWorker = Executors.newFixedThreadPool(threadCount, new ThreadFactoryBuilder().setNameFormat(threadName+"-%d").build());
+    }
+
+    public <E, T> void addTask(E event, Class<T> clazz) {
+        fsmWorker.submit(() -> {
+            if (clazz.getName().equals(LogReplicationRuntimeEvent.class.getName())) {
+                addLrRuntimeTask((LogReplicationRuntimeEvent) event);
```

Comment:
We are not sharing the thread pool between the FSMs. This class would be instantiated by both runtime and the replication FSMs.
I created a single class to avoid code duplication...please let me know if I have lowered readability in the process!! 
Added a class level comment in hopes of clarifying the intent.